typescript/web/src/components/app-lifecycle-manager/welcome-modal/welcome-modal.tsx

```diff
@@ -1,185 +1,357 @@
+      }
+      imagesAggregates {
+        totalCount
+      }
+      labelsAggregates {
+        totalCount
+      }
+      labelClassesAggregates {
+        totalCount
+      }
+    }
+  }
+`;
+
+const performWelcomeWorkflow = async ({
+  router,
   isServiceWorkerActive,
+  setIsServiceWorkerActive,
+  client,
+  setParamModalWelcome,
+  setIsLoadingWorkerAndDemo,
+  setBrowserError,
 }: {
+  router: NextRouter;
   isServiceWorkerActive: boolean;
+  setIsServiceWorkerActive: (state: boolean) => void;
+  setParamModalWelcome: (
+    state: WelcomeModalParam,
+    updateType: "replaceIn"
+  ) => void;
+  setBrowserError: (state: Error) => void;
+  setIsLoadingWorkerAndDemo: (state: boolean) => void;
+  client: ApolloClient<{}>;
 }) => {
+  try {
+    if (isServiceWorkerActive) {
+      return;
+    }
+
+    setParamModalWelcome(undefined, "replaceIn");
+
+    setIsLoadingWorkerAndDemo(true);
+
+    try {
+      await checkServiceWorkerReady();
+    } catch (e) {
+      if (e.message === messageNoWindow) {
+        // In Next JS SSR, the window is not available.
+        return;
+      }
+      if (process.env.STORYBOOK) {
+        // In Storybook, the Service Worker is not available.
+        return;
+      }
+      throw e;
+    }
+    setIsServiceWorkerActive(true);
+
+    const { data: getDatasetsResult } = await client.query({
+      query: getDatasetsQuery,
+    });
+
+    const demoDataset =
+      getDatasetsResult?.datasets == null
+        ? undefined
+        : getDatasetsResult?.datasets.filter(
+            (dataset: DatasetType) => dataset.name === "Demo dataset"
+          )?.[0] ?? undefined;
+
+    if (!demoDataset) {
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      const { data: createDemoDatasetResult, errors: createDemoDatasetErrors } =
+        await client.mutate({
+          mutation: createDemoDatasetQuery,
+          refetchQueries: [
+            "getDatasets",
+            "getDatasetById",
+            "getAllImagesOfADataset",
+            "getDataset",
+            "countLabelsOfDataset",
+            "getLabelClassesOfDataset",
+            "image",
+            "getImageLabels",
+          ],
+        });
+      if (createDemoDatasetErrors) {
+        throw createDemoDatasetErrors[0];
+      }
+    }
+
+    await router.prefetch(tutorialDatasetFirstImageUrl);
+
+    setIsLoadingWorkerAndDemo(false);
+  } catch (error) {
+    setBrowserError(error);
+  }
+};
+
+export const WelcomeModal = ({
+  initialIsServiceWorkerActive = false,
+  initialBrowserWarning = false,
+  initialIsLoadingWorkerAndDemo = false,
+  initialBrowserError = undefined,
+}: {
+  initialIsServiceWorkerActive?: boolean;
+  initialBrowserWarning?: boolean;
+  initialIsLoadingWorkerAndDemo?: boolean;
+  initialBrowserError?: Error;
+}) => {
+  const router = useRouter();
+
+  const handleError = useErrorHandler();
+  const client = useApolloClient();
+
+  const [{ hasUserTriedApp }, setHasUserTriedApp] = useCookies([
```

Comment:
Nitpicking, feel free to ignore

I would rename `setHasUserTriedApp` into `setCookie`. 
Later you do `setHasUserTriedApp("hasUserTriedApp", "false");` while `setCookie("hasUserTriedApp", "false");` could be more accurate.

Reply:
I know this is weird, but there would be duplication, we have `setHasUserTriedApp` and `setTryDespiteBrowserWarning` which would BOTH be called `setCookie`. And they are not interchangeable, their types are not compatible. So don't really have a choice here.